File: IN5450/Project1/ASA.m
%
% Script that sets up and runs an anglular spectrum approach simulation
%

%%%%%%%%%%%%%%%%%%
% Parameters read from file

P = ASAparameters;

%%%%%%%%%%%%%%%%%%
% Source, i.e. U(x,y,z_0), defined in separate function

% U0 = ASAsource(P,'piston',15*P.lambda);

U0 = ASAsource(P,'square',15*P.lambda);

%%%%%%%%%%%%%%%%%%
% Define propagator for distance (z-z0)
z = 0.08; % [m]
Prop = ASApropagator(P,z);

U = ifft2t(fft2t(U0).*Prop);
%%
figure(1); imagesc(P.ax,P.ay,db(abs(U))); 
CA = caxis; caxis(CA(2) + [-40 0]); colorbar;
xlabel('x [m]'); ylabel('y [m]'); 
title(['Wave field at depth z = ',num2str(100*z,'%.1f'),' cm. Estimated using ASA.']);
style_plot(gcf(), 'nord');
exportgraphics(gcf(), "1a.pdf", "BackgroundColor", "none", "ContentType", "vector");

%%
figure(2);
plot(P.ax,db(abs(U(:,floor(P.Nx/2)))),'LineWidth',2);
ylim(max(db(abs(U(:))))+[-40 0]); grid on;
xlabel('x [m]'); ylabel('[dB]');
title(['Center cut of wave field at depth z = ',num2str(100*z,'%.1f'),' cm. Estimated using ASA.']);
style_plot(gcf(), 'nord');
exportgraphics(gcf(), "1b.pdf", "BackgroundColor", "none", "ContentType", "vector")



%% Apply Taper
Kernel = hanning(3)*hanning(3).';
Kernel = Kernel/sum(Kernel(:));
U0Mod = conv2(U0, Kernel, 'same');

%%
z = 0.40;
dz = 2*P.lambda;
nZ = round(z/dz);
Zs = linspace(0,z,nZ);
Resp = zeros(P.Nx, nZ);
Prop = ASApropagator(P, dz);
U = ifft2t(fft2t(U0Mod).*Prop);
Resp(:, 1) = U(:, floor(P.Ny/2)).';

for ii = 2:nZ
    if mod(ii, 25) == 0
        fprintf('#')
    end
    U = ifft2t(fft2t(U).*Prop);
    Resp(:, ii) = U(:,floor(P.Ny/2)).';
end
%% Oppgave 2
figure(3); 

imagesc(Zs, P.ax, db(abs(Resp)));
cbar = colorbar; cbar.Label.String = "Pressure [dB]";
clim(max(db(abs(Resp(:))))+[-40 0]);
xlabel("z [m]"); ylabel("x [m]");
title("Wave field for y = 0 along z. Estimated using ASA.")
style_plot(gcf(), "nord");
exportgraphics(gcf, "2a.pdf", "BackgroundColor", "none", "ContentType", "vector")


figure(4); 
subplot(3,1,1);
plot(Zs,db(abs(Resp(floor(P.Nx/2), :)))); 
xlabel("z [m]"); ylabel("Pressure [dB]");
grid on;
title("Pressure along z at x = 0.")

subplot(3,1,2);
nearfar_limit = 0.14;
plot(P.ax, db(abs(Resp(:, floor(nearfar_limit/dz) ))))
xlabel("x [m]"); ylabel("Pressure [dB]");
ylim(max(db(abs(Resp(:))))+[-40 0]); grid on;
title("Center-cut of wave field along x at near-field farfield limit (z=0.14)");

subplot(3,1,3);
plot(P.ax, db(abs(Resp(:, end))))
xlabel("x [m]"); ylabel("Pressure [dB]");
ylim(max(db(abs(Resp(:))))+[-40 0]); grid on;
title("Center-cut of wave field along x at end of simulation at z = 0.4m");
style_plot(gcf(), "nord");
exportgraphics(gcf, "2b.pdf", "BackgroundColor", "none", "ContentType", "vector")

File: IN5450/Project1/ASA.jl
using Revise
using DSP
using Plots
using Printf

default(background_color=:transparent, foreground_color=:white)

################
# Load the precode
include("precode.jl")

################
# Parameters
P = ParamStruct()

################
# Source, i.e. U(x,y,z_0)

U0 = ASAsource(P, "square", 15*P.lambda)

################
# Define propagator for distance (z-z0)
z = 0.08
Prop = ASApropagator(P, z)

U0 = ifft2t(fft2t(U0).*Prop)
kernel = hanning(5)*hanning(5)'
kernel = kernel/sum(kernel)
U0Mod = conv(U0, kernel)[3:end-2, 3:end-2]


################
# Plot some images or something
sig2db = amp2db ∘ abs

heatmap(
    P.ax, P.ay, 
    U0 .|> sig2db, #abs .|> amp2db,
    xlabel="x [m]",
    ylabel="y [m]",
    clims = (-36 ,3.7),
    colorbar_title=" \nResponse [dB]",
    top_margin=5Plots.mm,
    right_margin=8Plots.mm
)
title!(@sprintf "Wave field at depth z = %.1f cm.\n Estimated using ASA." 100*z)
savefig("1a.svg")

plot(
    P.ax, U0[:, P.Nx÷2] .|> sig2db,
    linewidth=2,
    ylim = maximum(U0 .|> sig2db) .+ (-40, 0),
    xlabel = "x [m]", ylabel="Response [dB]",
    legend = :none
)
title!(@sprintf "Center cut of wave field at depth z = %.1f cm.\n Estimated using ASA." 100*z)
savefig("1b.svg")

# 2?
z = 0.40
dz = 2*P.lambda
nZ = round(z / dz) |> Int
Zs = range(0,z,nZ)
Resp = Complex.(zeros(P.Nx, nZ))

Prop = ASApropagator(P, dz)
U = ifft2t(fft2t(U0Mod) .* Prop)
Resp[:, 1] .= U[:, P.Ny÷2]

for ii = 2:nZ
    if ii % 25 == 0
        print("#")
    end
    global U = ifft2t(fft2t(global U).*Prop)
    Resp[:, ii] .= U[:, P.Ny÷2]
end

Resp

heatmap(
    P.ax, Zs,
    Resp .|> sig2db,
    xlabel="x [m]", ylabel="z [m]",
    clims = (-36 ,3.7),
    colorbar_title=" \nResponse [dB]",
    top_margin=5Plots.mm,
    right_margin=8Plots.mm
)

size(Resp)

plot(
    P.ax, Resp[:, nZ÷2] .|> sig2db
)